src/AlgebraicGeometry/Surfaces/EnriquesAut.jl

```diff
@@ -0,0 +1,908 @@
+numerical lattices of ``Y`` and ``X``.
+
+Most importantly the type stores a chain of lattices
+
+```math
+S_Y(2) \subseteq S_X \subseteq L_{1,25}
+```
+
+where ``L_{1,15}`` is an even unimodular lattice of signature $(1,25)``. 
+```
+"""
+mutable struct EnriquesBorcherdsCtx
+  # SY < SX < L26
+  L26::ZZLat
+  SX::ZZLat # numerical lattice of X
+  SY::ZZLat # numerical lattice of Y with form rescaled by 2
+  # the following are given with respect to the basis of SY
+  initial_walls::Vector{ZZMatrix}
+  initial_rays::Vector{ZZMatrix}
+  initial_isotropic_rays::Vector{ZZMatrix}
+  # automorphisms of the initial chamber
+  # written w.r.t. the basis of SY
+  initial_automorphisms::Vector{ZZMatrix}
+  initial_automorphisms_mod2::Vector{FqMatrix}  # same order as above but mod 2
+  initial_chamber::K3Chamber   # a non-degenerate L26/SX chamber
+  # whether a -2 vector defines an outer wall is a mod 2 condition
+  roots_mod2::Set{FqMatrix}  # mod 2 images of splitting roots
+  gramSY::ZZMatrix  # to avoid coercions
+  gramSX::ZZMatrix  # to avoid coercions
+  membership_test
+  # needed for the membership test
+  # the following are in SY/2SY w.r.t the basis of SY mod 2
+  Dplus_perp::FqMatrix
+  Dplus::FqMatrix
+  imgs_mod2::Set{FqMatrix}
+  # w.r.t the basis given by Dplus
+  Gplus
+  Gplus_mat::MatrixGroup{FqFieldElem, FqMatrix}
+  function EnriquesBorcherdsCtx()
+    return new()
+  end
+
+  volume_index::ZZRingElem
+  orderGbar::ZZRingElem
+end
+
+function Base.show(io::IO, dat::EnriquesBorcherdsCtx)
+  io = pretty(io)
+  print(io, "Enriques Borcherds context with det(SX) = $(det(dat.SX)).")
+end
+
+@doc raw"""
+    EnriquesBorcherdsCtx(SY::ZZLat, SX::ZZLat, L26::ZZLat, weyl::ZZMatrix)
+
+# Input:
+-`SY` `SX` and `L26` must be an ascending chain of lattices in the same quadratic space. 
+- `weyl` -- a Weyl vector of `L26` given with respect to the basis of the lattice `L26`. 
+"""
+function EnriquesBorcherdsCtx(SY::ZZLat, SX::ZZLat, L26::ZZLat, weyl::ZZMatrix)
+  # X K3 ---> Y Enriques
+  ECtx = EnriquesBorcherdsCtx()
+  ECtx.L26 = L26
+  ECtx.SX = SX
+  ECtx.SY = SY
+  ECtx.gramSY = change_base_ring(ZZ, gram_matrix(SY))
+  ECtx.gramSX = change_base_ring(ZZ, gram_matrix(SX))
+
+  @vprint :EnriquesAuto 2 "computing Borcherds context\n"
+  dataY,_ = BorcherdsCtx(L26, SY, weyl; compute_OR=false)
+  dataY.membership_test = (x -> true)
+  ECtx.initial_chamber = chamber(dataY, dataY.weyl_vector; check=true)
+  # SY + Sm < SX is a primitive extension with glue map phi: D(Sm) -> D(SY)
+  Sm = orthogonal_submodule(SX, SY)
+  phi, inc_Dminus, inc_Dplus = glue_map(SX, Sm, SY)
+  # H_Sm = pi_Sm(SX) note that H_Sm/Sm is
+  H_Sm = cover(domain(phi))
+  sv2 = [1//2*(i[1]*basis_matrix(Sm)) for i in short_vectors(Sm, 4) if i[2] == 4]
+  sv2 = [domain(phi)(i) for i in sv2 if i in H_Sm]
+  ECtx.roots_mod2 = Set([change_base_ring(GF(2), solve(basis_matrix(SY), matrix(QQ, 1, 26, 2*lift(phi(i))); side=:left)) for i in sv2])
```

Comment:
```suggestion
  ECtx.roots_mod2 = Set(change_base_ring(GF(2), solve(basis_matrix(SY), matrix(QQ, 1, 26, 2*lift(phi(i))); side=:left)) for i in sv2)
```
this saves some allocations